File: challenges/6/solution.py
def convert(s: str, numRows: int) -> str:
    # scorrere la string num_rows volte
    # ogni volta ti sposti avanti di (num_rows-1)*2
    idx = 0
    row = 0
    total_step = (numRows-1)*2
    is_first_step = True

    # edge cases
    output = s[0]
    if len(s) == 1:
        return output
    if len(s) < numRows:
        return s
    if numRows == 1:
        return s


    while(True):
        
        first_step = (numRows-1-row)*2
        second_step = total_step - first_step
        do_step = False

        # first step
        if is_first_step:
            is_first_step = False
            if first_step != 0:
                do_step = True
                idx += first_step
    

        # second step
        else:
            is_first_step = True
            if second_step != 0:
                do_step = True
                idx += second_step


        # restart with new line
        if idx >= len(s):

            row += 1

            if row >= numRows:
                break
            
            # reset index 
            is_first_step = True
            idx = row
            output += s[idx] # add first 
            continue
        
    
        # add char to the string
        if do_step:
            output += s[idx]

    return output

File: challenges/6/solution.js
/**
 * @param {string} s
 * @param {number} idx
 * @return {string}
 */
const isInsideBoundaries = (s, idx) => {
  return idx < s.length;
};

/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
const convert = (s, numRows) => {
  if (s.length <= numRows) return s;
  if (numRows < 2) return s;

  let currIndex = 0;
  let output = "";
  const step = (numRows - 1) * 2;

  for (let i = 0; i < numRows; i++) {
    const rowStepDiagonal = (numRows - 1 - i) * 2;
    const rowStepVertical = step - rowStepDiagonal;
    currIndex = i;
    output += s[currIndex];

    console.log(rowStepDiagonal, rowStepVertical);

    while (isInsideBoundaries(s, currIndex)) {
      currIndex += rowStepDiagonal;
      if (isInsideBoundaries(s, currIndex) && rowStepDiagonal != 0)
        output += s[currIndex];
      currIndex += rowStepVertical;
      if (isInsideBoundaries(s, currIndex) && rowStepVertical != 0)
        output += s[currIndex];
    }
  }

  return output;
};
